### PR TITLE
Add the sqlite3 with knex to store db

### PR DESCRIPTION
**Pull request by thuongmai**
Create a framework for storing database
- [x] Create Register and Login logic
- [x] Hashing the password and validate email to avoid injecting malicious code (SQL or XSS attack)
- [x] Return a JWT token for user session

### CODE REVIEW

File: back-end/routes/authApi.js [hunk 1]
```diff
@@ -5,6 +5,6 @@ const UserController = require('../controllers/UserController');
 // Login route
 router.post('/login', UserController.login);
 router.post('/register', UserController.register);
-router.post('/logout', UserController.logout);
+router.delete('/logout', UserController.logout);
```
Comment (chanhthi1502) [inline back-end/routes/authApi.js]: can remove this method, log out feature will be handled by front end

File: back-end/routes/taskManagementApi.js [hunk 2]
```diff
@@ -3,9 +3,9 @@ const router = express.Router();
...
 
 // Login route
-router.post('/create', TaskController.create);
-router.post('/read', TaskController.read);
-router.post('/update', TaskController.update);
-router.post('/delete', TaskController.delete);
+router.post('/tasks', TaskController.create); // create
+router.get('/tasks', TaskController.read); // read
+router.post('/tasks/:id', TaskController.update); // update
+router.delete('/delete/:id', TaskController.delete); // delete
```
Comment (chanhthi1502) [inline back-end/routes/taskManagementApi.js:9]: can remove this row as well

File: back-end/models/User.js [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+	* - When hashing the password, `bcrypt` library also concat the `salt`
+	* into the `hashedPassword` so when we are using the function `bycrypt.compare`,
+	* we no longer need to include `salt`. That is why we don't store salt in the db
+	*/
+	return await bcrypt.compare(password, hashedPassword);
+}
+
+module.exports = {
+	async findAccount(opt) {
+		if (_.isEmpty(opt)) {
```
Comment (chanhthi1502) [inline back-end/models/User.js:26]: please return error message as constant variable. For example: EMAIL_INVALID

File: back-end/db/knexfile.js [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    connection: {
+      filename: path.join(__dirname, 'db.sqlite3')
+    },
+    migrations: {
+      tableName: 'knex_migrations'
+    },
+    useNullAsDefault: true
+  },
+
+  // staging: {
```
Comment (chanhthi1502) [inline back-end/db/knexfile.js]: can we remove these comments if we don't need it ?

File: back-end/routes/taskManagementApi.js (hunk 2 above)
Comment (thuongmai) [reply back-end/routes/taskManagementApi.js:9]: Do we not want to delete the existing task ?

Comment (thuongmai) [reply back-end/routes/taskManagementApi.js:9]: HEY

File: back-end/controllers/UserController.js [hunk 5]
```diff
@@ -1,25 +1,72 @@
...
 
-		// Verify password
+	async register(req, res) {
+		try {
+			const userFound = await User.findAccount(req.body);
+			if (!userFound) {
+				const registerStatus = await User.registerUser(req.body);
 
-		// Generate JWT
+				// if (registerStatus.emailAlreadyExists) {
```
Comment (chanhthi1502) [inline back-end/controllers/UserController.js]: Please remove comment here

File: back-end/controllers/UserController.js [hunk 6]
```diff
@@ -1,25 +1,72 @@
...
-		res.status(405).json({ message: 'Login is not supported via POST requests' });
+				if (registerStatus) {
+					const sessionToken = jwt.sign(
+						{ email: req.body.email },
+						globalUtil.jwt_secret_key,
+						{ expiresIn: globalUtil.session_expired });
+					return res.status(200).json({
+						sessionToken,
+						registerStatus: true,
+						message: 'Registered done'
```
Comment (chanhthi1502) [inline back-end/controllers/UserController.js]: Can we make message constant ? For example const REGISTER_DONE_MESSAGE = 'Registered done'

File: back-end/routes/taskManagementApi.js [hunk 7]
```diff
@@ -3,9 +3,9 @@ const router = express.Router();
 const TaskController = require('../controllers/TaskController');
 
 // Login route
-router.post('/create', TaskController.create);
-router.post('/read', TaskController.read);
-router.post('/update', TaskController.update);
-router.post('/delete', TaskController.delete);
+router.post('/tasks', TaskController.create); // create
+router.get('/tasks', TaskController.read); // read
+router.post('/tasks/:id', TaskController.update); // update
```
Comment (chanhthi1502) [inline back-end/routes/taskManagementApi.js:8]: in update feature , can we update by username, email ? or we just can only update by id ?

File: back-end/controllers/UserController.js [hunk 8]
```diff
@@ -1,25 +1,72 @@
...
+					});
+				}
+				return res.status(401).json({ message: 'Registered failed' });
+			}
+			return res.status(401).json({ message: 'Account already exists' });
+		} catch (error) {
+			console.error('Failed to register:', error);
+			return res.status(405).json({ message: `Failed to register new user. ${error}` });
```
Comment (chanhthi1502) [inline back-end/controllers/UserController.js]: According to Chat GPT , we should return status 400 when Failed to register new user haha

**Review by chanhthi1502 — CHANGES_REQUESTED** [review]

File: back-end/routes/taskManagementApi.js (hunk 7 above)
Comment (thuongmai) [reply back-end/routes/taskManagementApi.js:8]: This ID is the task ID, not user ID.

Each user will have multiple tasks and each task will have a unique task ID. So we need the task ID from the request handler to update a task in the db. What scenario that you want to update a task by username/email ?

File: back-end/controllers/UserController.js (hunk 8 above)
Comment (thuongmai) [reply back-end/controllers/UserController.js]: Good catch 😆 
By the defintion:
- The 405 error happens when a client attempts to use an HTTP method that is not allowed `by the server`
 - The 400 error happens when the server cannot or will not process the request due to something that is perceived to be a `client error`

So this is partly my fault for not taking care of all the errors when registering the user. My initial thought is to let `UserController#register` to validate variables that user input, and the `models/User#registerUser` will handle server error. 

With that being said, the `try/catch` in the `UserController#register` will actually throw error from the server and it should be 405 (a.k.a register new record in the database). 

We can log the ticket for this to go back into it once we develop a throughout system. What do you think ?

**Pull request merged**